### DISCUSSION TITLE
NCP and CLI

### DISCUSSION THREAD

**Discussion by dheryta** [openthread/openthread#8347] [Q&A] [answered]
I have 2 NRF52840-MDK chips configured in NCP mode and other in CLI mode.
They are not able to ping each other.
Their mesh prefix are different.
I tried to make them same with config-gateway and add-prefix commands.
But it does not seem to help.
Any help is greatly appreciated!

**Comment by jwhui** [comment, marked as answer]
Did you try the [Thread Border Router codelab](https://openthread.io/codelabs/openthread-border-router)? If you want the latest features, I suggest RCP mode (instead of NCP mode).

**Comment by dheryta** [comment]
1. I don’t need internet access or access to other IP networks, do I still need a border router? Pinging link local address works fine but mesh local doesn’t, which is kind of obvious too, given their prefix are different routing us not working the v way it should.
2. I was planning to use  iperf with NCP mode to do some benchmarking
tests. I have not worked with RCP, Will it allow using iperf? If not what
options do we have for benchmarking tests?

On Mon, Oct 31, 2022 at 11:39 PM Jonathan Hui ***@***.***>
wrote:

> Did you try the Thread Border Router codelab
> <https://openthread.io/codelabs/openthread-border-router>? If you want
> the latest features, I suggest taking a look at RCP mode (instead of NCP
> mode).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openthread/openthread/discussions/8347#discussioncomment-4024227>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLSMKOEGCUFHGL5YNJFRWLWGCNHFANCNFSM6AAAAAARTWN3PA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
--
Regards

Dheryta

**Reply by jwhui** [reply]
> 1. I don’t need internet access or access to other IP networks, do I still need a border router? Pinging link local address works fine but mesh local doesn’t, which is kind of obvious too, given their prefix are different routing us not working the v way it should.

You don't strictly need a Thread Border Router, but the codelab is relatively straightforward. The codelab also walks you through the proper way to configure a Thread network that avoids the issues you are running into.

> 2. I was planning to use  iperf with NCP mode to do some benchmarking tests. I have not worked with RCP, Will it allow using iperf? If not what options do we have for benchmarking tests?

Yes, you can use iperf. On OTBR, the Thread network is simply exposed as a standard network interface.

**Comment by dheryta** [comment]
Thanks, it works the way I needed.
Now, what do we do if I want these nodes to auto join the existing thread
network whenever they power up?


--
Regards

Dheryta

**Reply by jwhui** [reply]
The CLI makes direct calls to the OpenThread APIs, you can use the CLI implementation as an example of how to call the OpenThread APIs. To have the device attach automatically after power up, you can update the [examples/apps/cli/main.c](https://github.com/openthread/openthread/blob/main/examples/apps/cli/main.c) to perform the equivalent of `ifconfig up` and `thread start`.

**Reply by dheryta** [reply]
I am running into an issue, I am not sure why.
I setup Border Router, 1 RCP node, 1 CLI node.
Network is started with Border Router.
I set network key in RCP node and CLI node
Both nodes join the network but their mesh local prefix does not match.
It is different for all 3 nodes.
I have not done SRP part in border router codelab.

**Reply by jwhui** [reply]
Can you provide more detail so that someone can reproduce the issue? Specifically, it would help if you provided the exact commands you are using to form the network on the border router and how you are setting the key on the CLI node.

**Reply by dheryta** [reply]
I followed the codelab - https://openthread.io/codelabs/openthread-border-router#0 to create border router.
On the RPi I used following commands to create the network
ot-ctl dataset
ot-ctl dataset networkname TestNetwork
ot-ctl dataset networkkey 1234C0DE1AB51234C0DE1AB51234C0DE
ot-ctl dataset channel 15
ot-ctl dataset panid 0x2222
ot-ctl dataset extpanid C0DE1AB5C0DE1AB5
ot-ctl dataset commit active
ot-ctl ifconfig up
ot-ctl thread start

RCP Node is created with codelab - https://openthread.io/codelabs/openthread-hardware#3
With ot-ctl commands I set networkkey here with 
sudo ./ot-ctl dataset networkkey 1234C0DE1AB51234C0DE1AB51234C0DE
sudo ./ot-ctl thread start

CLI Node is created with commands script/build nrf52840 USB_trans -DOT_MTD=OFF -DOT_APP_RCP=OFF -DOT_RCP=OFF and modified main.c where I set NetworkName and NetworkKey with APIs, following method added
void setNetworkConfiguration(otInstance *aInstance)
{
    static char          aNetworkName[] = "TestNetwork";
    otOperationalDataset aDataset;

    memset(&aDataset, 0, sizeof(otOperationalDataset));

    /*
     * Fields that can be configured in otOperationDataset to override defaults:
     *     Network Name, Mesh Local Prefix, Extended PAN ID, PAN ID, Delay Timer,
     *     Channel, Channel Mask Page 0, Network Key, PSKc, Security Policy
     */
    aDataset.mActiveTimestamp.mSeconds             = 1;
    aDataset.mActiveTimestamp.mTicks               = 0;
    aDataset.mActiveTimestamp.mAuthoritative       = false;
    aDataset.mComponents.mIsActiveTimestampPresent = true;

    /* Set Channel to 15 */
    aDataset.mChannel                      = 15;
    aDataset.mComponents.mIsChannelPresent = true;

    /* Set Pan ID to 2222 */
    aDataset.mPanId                      = (otPanId)0x2222;
    aDataset.mComponents.mIsPanIdPresent = true;

    /* Set Extended Pan ID to C0DE1AB5C0DE1AB5 */
    uint8_t extPanId[OT_EXT_PAN_ID_SIZE] = {0xC0, 0xDE, 0x1A, 0xB5, 0xC0, 0xDE, 0x1A, 0xB5};
    memcpy(aDataset.mExtendedPanId.m8, extPanId, sizeof(aDataset.mExtendedPanId));
    aDataset.mComponents.mIsExtendedPanIdPresent = true;

    /* Set network key to 1234C0DE1AB51234C0DE1AB51234C0DE */
    uint8_t key[OT_NETWORK_KEY_SIZE] = {0x12, 0x34, 0xC0, 0xDE, 0x1A, 0xB5, 0x12, 0x34, 0xC0, 0xDE, 0x1A, 0xB5, 0x12, 0x34, 0xC0, 0xDE};
    memcpy(aDataset.mNetworkKey.m8, key, sizeof(aDataset.mNetworkKey));
    aDataset.mComponents.mIsNetworkKeyPresent = true;

    /* Set Network Name to TestNetwork */
    size_t length = strlen(aNetworkName);
    assert(length <= OT_NETWORK_NAME_MAX_SIZE);
    memcpy(aDataset.mNetworkName.m8, aNetworkName, length);
    aDataset.mComponents.mIsNetworkNamePresent = true;

    otDatasetSetActive(aInstance, &aDataset);
    /* Set the router selection jitter to override the 2 minute default.
       CLI cmd > routerselectionjitter 20
       Warning: For demo purposes only - not to be used in a real product */
    //uint8_t jitterValue = 20;
    //otThreadSetRouterSelectionJitter(aInstance, jitterValue);
}

**Reply by jwhui** [reply]
> I followed the codelab - https://openthread.io/codelabs/openthread-border-router#0 to create border router.
On the RPi I used following commands to create the network
ot-ctl dataset
ot-ctl dataset networkname TestNetwork
ot-ctl dataset networkkey 1234C0DE1AB51234C0DE1AB51234C0DE
ot-ctl dataset channel 15
ot-ctl dataset panid 0x2222
ot-ctl dataset extpanid C0DE1AB5C0DE1AB5
ot-ctl dataset commit active
ot-ctl ifconfig up
ot-ctl thread start

The commands you used above to form the Thread network are not the same as in the codelab. Please follow the instructions in: https://openthread.io/codelabs/openthread-border-router#2

**Reply by dheryta** [reply]
If I follow the codelab and execute following commands
$ sudo ot-ctl dataset init new
Done
$ sudo ot-ctl dataset commit active
Done
$ sudo ot-ctl ifconfig up
Done
$ sudo ot-ctl thread start
Done

then, the network settings keep getting changed. I need static name for network, key, channel, panid, extpanid so that my cli nodes can connect to the network on boot.

Also, I am unable to understand why this customization would affect the thread mesh IP addresses?

**Reply by jwhui** [reply]
It is important to start with a complete operational dataset. The following command will create one for you:
```
sudo ot-ctl dataset init new
```
It is possible to customize individual parameters before the `dataset commit active` command. So you can set the other values as you like.

**Reply by dheryta** [reply]
Got it.
I started with dataset init new.
and followed earlier steps for RCP where I set network key and thread start - it creates its own network
and CLI chip is joining the border router network but it has a different mesh prefix.

**Reply by jwhui** [reply]
On the CLI node, only set the network key. Do not set any of the other operational dataset parameters.

The problem you had in both cases is that you were configuring a partial operational dataset. You should either configure the entire dataset or none of it. Anything in between, you will run into the issues you observe.

**Reply by dheryta** [reply]
All is working now.
I am now working on benchmarking.
I tried iperf and iperf3. 
With iperf, UDP works fine but for TCP, connection is refused. With iperf3 both TCP and UDP are not working, connection is refused.
While using iperf I also set bandwidth less than 250K and packet size < 1232 with UDP.

I also tried the tcp benchmark as per https://github.com/openthread/openthread/blob/main/src/cli/README_TCP.md#benchmark-size but I keep running into error - "TCP Benchmark Failed". I am unable to send hello packet as well.

Do I need to set any flag for TCP to work or I am doing something wrong?

**Reply by jwhui** [reply]
If you are running OpenThread on Linux, you'll want to disable the TCP feature to avoid conflicting with Linux's TCP stack. OTBR disables this by default: https://github.com/openthread/ot-br-posix/blob/22bcc573ebb99ea0695ab73dc83ff36081d95935/third_party/openthread/CMakeLists.txt#L102
